core/common/src/main/java/alluxio/URI.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the “License”). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio;
+
+import alluxio.collections.Pair;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Serializable;
+import java.net.URISyntaxException;
+
+/**
+ * This interface represents a URI. This {@link URI} supports more than just strict
+ * {@link java.net.URI}. Some examples:
+ *   * Windows paths
+ *     * C:\
+ *     * D:\path\to\file
+ *     * E:\path\to\skip\..\file
+ *   * URI with multiple scheme components
+ *     * scheme://host:123/path
+ *     * scheme:part2://host:123/path
+ *     * scheme:part2:part3://host:123/path
+ *
+ * Currently, does not support fragment in the URI.
+ */
+public interface URI extends Comparable<URI>, Serializable {
```

Comment:
@calvinjia That is a good question. Well, currently, nothing other than our AlluxioURI uses this interface, and I don't foresee anything other than AlluxioURI to use it. What do you think it should be renamed to?